code/sha3/Hacl.Hash.SHA3.Simd256.fst

```diff
@@ -0,0 +1,353 @@
+               (input1 <: lbuffer uint8 inputByteLen)
+               (input2 <: lbuffer uint8 inputByteLen)
+               (input3 <: lbuffer uint8 inputByteLen) /\
+       live4 h output0 output1 output2 output3 /\
+       internally_disjoint4 output0 output1 output2 output3)
+     (ensures  fun h0 _ h1 ->
+       modifies (loc output0 |+| loc output1 |+| loc output2 |+| loc output3) h0 h1 /\
+       as_seq h1 output0 == S.sha3_384 (v inputByteLen) (as_seq h0 (input0 <: lbuffer uint8 inputByteLen)) /\
+       as_seq h1 output1 == S.sha3_384 (v inputByteLen) (as_seq h0 (input1 <: lbuffer uint8 inputByteLen)) /\
+       as_seq h1 output2 == S.sha3_384 (v inputByteLen) (as_seq h0 (input2 <: lbuffer uint8 inputByteLen)) /\
+       as_seq h1 output3 == S.sha3_384 (v inputByteLen) (as_seq h0 (input3 <: lbuffer uint8 inputByteLen)))
+let sha3_384 output0 output1 output2 output3
+        input0 input1 input2 input3 inputByteLen =
+  admit();
+  let ib = ntup4 (input0,(input1,(input2,input3))) in
+  let rb = ntup4 (output0,(output1,(output2,output3))) in
+  keccak #SHA3_384 #M256 832ul (* 768ul *) inputByteLen ib (byte 0x06) 48ul rb
+
+val sha3_512:
+  output0:lbuffer uint8 64ul
+  -> output1:lbuffer uint8 64ul
+  -> output2:lbuffer uint8 64ul
+  -> output3:lbuffer uint8 64ul
+  -> input0:buffer_t MUT uint8
+  -> input1:buffer_t MUT uint8
+  -> input2:buffer_t MUT uint8
+  -> input3:buffer_t MUT uint8
+  -> inputByteLen:size_t{v inputByteLen == length input0 /\ 
+                         v inputByteLen == length input1 /\
+                         v inputByteLen == length input2 /\
+                         v inputByteLen == length input3}
+  -> Stack unit
+     (requires fun h ->
+       live4 h (input0 <: lbuffer uint8 inputByteLen)
+               (input1 <: lbuffer uint8 inputByteLen)
+               (input2 <: lbuffer uint8 inputByteLen)
+               (input3 <: lbuffer uint8 inputByteLen) /\
+       live4 h output0 output1 output2 output3 /\
+       internally_disjoint4 output0 output1 output2 output3)
+     (ensures  fun h0 _ h1 ->
+       modifies (loc output0 |+| loc output1 |+| loc output2 |+| loc output3) h0 h1 /\
+       as_seq h1 output0 == S.sha3_512 (v inputByteLen) (as_seq h0 (input0 <: lbuffer uint8 inputByteLen)) /\
+       as_seq h1 output1 == S.sha3_512 (v inputByteLen) (as_seq h0 (input1 <: lbuffer uint8 inputByteLen)) /\
+       as_seq h1 output2 == S.sha3_512 (v inputByteLen) (as_seq h0 (input2 <: lbuffer uint8 inputByteLen)) /\
+       as_seq h1 output3 == S.sha3_512 (v inputByteLen) (as_seq h0 (input3 <: lbuffer uint8 inputByteLen)))
+let sha3_512 output0 output1 output2 output3
+        input0 input1 input2 input3 inputByteLen =
+  admit();
+  let ib = ntup4 (input0,(input1,(input2,input3))) in
+  let rb = ntup4 (output0,(output1,(output2,output3))) in
+  keccak #SHA3_512 #M256 576ul (* 1024ul *) inputByteLen ib (byte 0x06) 64ul rb
+
+val state_malloc:
+    r:rid
+  -> ST.ST (s:buffer uint64 { length s = 100 })
```

Comment:
use lbuffer?

why is this helper needed?

Reply:
Using lbuffer will throw `(Error 331) This name is being ignored` and `(Error 19) Subtyping check failed`. Removing the helper will throw the first error.

Reply:
what syntax are you using? I am thinking
```
val state_malloc: r:rid -> ST.ST (lbuffer uint64 100)
```

(or whatever the order of the arguments to lbuffer is)

Reply:
I tried `val state_malloc: r:rid -> ST.ST (lbuffer uint64 100ul)`, it throws `(Error 331) This name is being ignored`

Reply:
but which name is being ignored? you should have a precise location for the error

Reply:
also, I'm wondering if adding a type ascription at call-site, such as:
```
let b: lbuffer uint64 100 = malloc ...
```
will help at al